File: semantic-security-101/semantic_security_101.py
#!/usr/bin/env python3

import socket
import base64

def main():
    # connect 
    s = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    host = '195.134.67.7'
    port = 10090

    s.connect((host,port))
    s = s.makefile(mode="rw")

    # print welcome messages
    welcome1 = s.readline().strip()
    welcome2 = s.readline().strip()
    welcome3 = s.readline().strip()
    print(welcome1,"\n",welcome2,"\n",welcome3,"\n")

    for i in range(10):
        # t0 will be used twice, while the other two just once
        t0 = 'A' * 16
        t1 = 'B' * 16
        t2 = 'C' * 16

        cmd = f"encrypt({t0},{t1})"
        print(cmd,file=s,flush=True)
    
        ct = s.readline().strip()
        
        # get just the ciphertext
        parts = ct.split(': ')
        
        if( i == 0):
            ct0 = parts[1].strip()
        else: # avoid the score part(just print it)
            print(parts[0].strip())
            ct0 = parts[2].strip()
        print("cipher:",ct0)

        # do the same but for message t0 and t2
        cmd = f"encrypt({t0},{t2})"
        print(cmd,file=s,flush=True)

        ct = s.readline().strip()

        parts = ct.split(': ')
        ct1 = parts[1].strip()
        print("cipher:",ct1)
    
        if ct0 == ct1:  # if they are the same it means that the message t0 was encrypted
            solve_cmd = f"solve(0)"
        else:
            solve_cmd = f"solve(1)"

        print("solve_cmd",solve_cmd)
        print(solve_cmd, file=s, flush=True)

    # read the flag           
    flag = s.readline().strip()
    print(flag)
    s.close()

if __name__ == "__main__":
    main()

File: semantic-security-101/semantic_security_101.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strings"
)

func main() {
	// Connect to the server
	host := "195.134.67.7"
	port := "10090"
	conn, err := net.Dial("tcp", host+":"+port)
	if err != nil {
		fmt.Println("Error connecting to server:", err)
		return
	}
	defer conn.Close()

	reader := bufio.NewReader(conn)
	writer := bufio.NewWriter(conn)

	// Print welcome messages
	welcome1, _ := reader.ReadString('\n')
	welcome2, _ := reader.ReadString('\n')
	welcome3, _ := reader.ReadString('\n')
	fmt.Println(welcome1, "\n", welcome2, "\n", welcome3, "\n")

	for i := 0; i < 10; i++ {
		// Define the plaintexts
		t0 := strings.Repeat("A", 16)
		t1 := strings.Repeat("B", 16)
		t2 := strings.Repeat("C", 16)

		// Encrypt t0 and t1
		cmd := fmt.Sprintf("encrypt(%s,%s)\n", t0, t1)
		writer.WriteString(cmd)
		writer.Flush()

		ct, _ := reader.ReadString('\n')
		parts := strings.Split(strings.TrimSpace(ct), ": ")

		var ct0 string
		if i == 0 {
			ct0 = strings.TrimSpace(parts[2])
		} else {
			fmt.Println(parts[0])
			ct0 = strings.TrimSpace(parts[3])
		}
		fmt.Println("cipher:", ct0)

		// Encrypt t0 and t2
		cmd = fmt.Sprintf("encrypt(%s,%s)\n", t0, t2)
		writer.WriteString(cmd)
		writer.Flush()

		ct, _ = reader.ReadString('\n')
		parts = strings.Split(strings.TrimSpace(ct), ": ")
		
		ct1 := strings.TrimSpace(parts[2])
		fmt.Println("cipher:", ct1)

		// Compare ciphertexts and solve
		var solveCmd string
		if ct0 == ct1 {
			solveCmd = "solve(0)\n"
		} else {
			solveCmd = "solve(1)\n"
		}
		fmt.Println("solve_cmd", solveCmd)
		writer.WriteString(solveCmd)
		writer.Flush()
	}

	// Read the flag
	flag, _ := reader.ReadString('\n')
	fmt.Println(flag)
}

